src/System.Device.Gpio/System/Device/Gpio/Drivers/LibGpiodDriver.Linux.cs

```diff
@@ -0,0 +1,276 @@
+
+        protected internal override PinMode GetPinMode(int pinNumber)
+        {
+            if (_pinNumberToSafeLineHandle.TryGetValue(pinNumber, out SafeLineHandle pin))
+            {
+                return (Interop.GetLineDirection(pin) == 1) ? PinMode.Input : PinMode.Output;
+            }
+            return PinMode.Input;
+        }
+
+        protected internal override bool IsPinModeSupported(int pinNumber, PinMode mode)
+        {
+            // Libgpiod Api seems do not support pull up or pull down resistors.
+            return mode != PinMode.InputPullDown && mode != PinMode.InputPullUp;
+        }
+
+        protected internal override void OpenPin(int pinNumber)
+        {
+            SafeLineHandle pin = Interop.GetChipLineByOffset(_chip, pinNumber);
+            if (pin == null)
+            {
+                throw new IOException($"Pin number {pinNumber} not available for chip: {_chip} error: {Marshal.GetLastWin32Error()}");
+            }
+            _pinNumberToSafeLineHandle.Add(pinNumber, pin);
+        }
+
+        protected internal override PinValue Read(int pinNumber)
+        {
+            if (_pinNumberToSafeLineHandle.TryGetValue(pinNumber, out SafeLineHandle pin))
+            {
+                int value = Interop.GetGpiodLineValue(pin);
+                if (value == -1)
+                {
+                    throw new IOException($"Error while reading value from pin number: {pinNumber}, error: {Marshal.GetLastWin32Error()}");
+                }
+                return (value == 1) ? PinValue.High : PinValue.Low;
+            }
+            return PinValue.Low;
+        }
+
+        protected internal override void RemoveCallbackForPinValueChangedEvent(int pinNumber, PinChangeEventHandler callback)
+        {
+            if (_pinNumberToEventHandler.TryGetValue(pinNumber, out LibGpiodDriverEventHandler eventHandler))
+            {
+                eventHandler.ValueFalling -= callback;
+                eventHandler.ValueRising -= callback;
+                if (eventHandler.IsCallbackListEmpty())
+                {
+                    eventHandler.CancellationTokenSource.Cancel();
+                    Interop.ReleaseGpiodLine(eventHandler.PinHandle);
+                    eventHandler.Dispose();
+                    _pinNumberToEventHandler.TryRemove(pinNumber, out eventHandler);
+                 }
+            }
+            else
+            {
+                throw new InvalidOperationException("Attempted to remove a callback for a pin that is not listening for events.");
+            }
+        }
+
+        protected internal override void SetPinMode(int pinNumber, PinMode mode)
+        {
+            int success = -1;
+            if (_pinNumberToSafeLineHandle.TryGetValue(pinNumber, out SafeLineHandle pin))
+            {
+                if (mode == PinMode.Input)
+                {
+                    success = Interop.RequestLineInput(pin, pinNumber.ToString());
```

Comment:
NIT: why do some APIs take strings instead of numbers in our Interop library?

Reply:
2nd argument is just consumer name, as long as we are not using consumer name for a pin just putting pin number.